airflow/traces/otel_tracer.py

```diff
@@ -0,0 +1,316 @@
+
+        if start_time is not None:
+            start_time = datetime_to_nano(start_time)
+
+        if parent_sc is not None:
+            ctx = trace.set_span_in_context(NonRecordingSpan(parent_sc))
+            span = tracer.start_as_current_span(
+                span_name, context=ctx, attributes=attributes, links=_links, start_time=start_time
+            )
+        else:
+            span = tracer.start_as_current_span(
+                span_name, attributes=attributes, links=_links, start_time=start_time
+            )
+        return span
+
+    def start_span_from_dagrun(
+        self, dagrun, span_name: str | None = None, component: str = "dagrun", links=None
+    ):
+        """Produce a span from dag run."""
+        # check if dagrun has configs
+        conf = dagrun.conf
+        trace_id = int(gen_trace_id(dag_run=dagrun, as_int=True))
+        span_id = int(gen_dag_span_id(dag_run=dagrun, as_int=True))
+
+        if conf is not None:
+            traceparent = conf.get(TRACEPARENT)
+            tracestate = conf.get(TRACESTATE)
+
+        tracer = self.get_tracer(component=component, span_id=span_id, trace_id=trace_id)
+
+        if self.tag_string and tracestate:
+            tag_string = self.tag_string + "," + tracestate
+        else:
+            tag_string = self.tag_string or tracestate
+
+        if span_name is None:
+            span_name = dagrun.dag_id
+
+        if links is not None:
+            _links = gen_links_from_kv_list(links)
+        else:
+            _links = []
+
+        _links.append(
+            Link(
+                context=trace.get_current_span().get_span_context(),
+                attributes={"meta.annotation_type": "link", "from": "parenttrace"},
+            )
+        )
+
+        if traceparent is not None:
+            # add the trace parent as the link
+            _links.append(gen_link_from_traceparent(traceparent))
+
+        span_ctx = SpanContext(
+            trace_id=INVALID_TRACE_ID, span_id=INVALID_SPAN_ID, is_remote=True, trace_flags=TraceFlags(0x01)
+        )
+        ctx = trace.set_span_in_context(NonRecordingSpan(span_ctx))
+        span = tracer.start_as_current_span(
+            name=span_name,
+            context=ctx,
+            links=_links,
+            start_time=datetime_to_nano(dagrun.queued_at),
+            attributes=parse_tracestate(tag_string),
+        )
+        return span
+
+    def start_span_from_taskinstance(
+        self,
+        ti,
+        span_name: str | None = None,
+        component: str = "taskinstance",
+        child: bool = False,
+        links=None,
+    ):
+        """
+        Create and start span from given task instance.
+
+        Essentially the span represents the ti itself if child == True, it will create a 'child' span under the given span.
+        """
+        dagrun = ti.dag_run
+        trace_id = int(gen_trace_id(dag_run=dagrun, as_int=True))
+        span_id = int(gen_span_id(ti=ti, as_int=True))
```

Comment:
I hate to say it, but if we don't come up with a way to not have to cast this each time, then I feel like adding the as_int was a mistake.  Could you remind me why this has to be re-cast here?

Reply:
Sure, it's because the trace_id and span_id that was supposed to be fed into the Otel's tracer method required those types to be of 'int' types, and gen_trace_id( ... ) signature returns int | str so, hence the typing error.

Reply:
Right.   So even though we are casting it to an int before returning, the typechecking sees the return hints and thinks it could be either one.   hmmmmmmmmmm

Leave this comment open for me, I'll see if I can get some clever ideas over the weekend.

Reply:
Ok, will keep this open. Thanks!

Reply:
I think the as_int part needs to be reverted.  Sorry for the extra work.  I thought it would be clever and make the code cleaner, but MyPy forcing the casting to int() here is just making it more confusing I think.